### PR TITLE
Support scientific notation in `BigDecimal#to_s`

### LINKED ISSUE #9578: RFC: change default string representation in BigDecimal
Proposal: The default string representation for `BigDecimal`s should be in E (scientific) notation,

ie `BigDecimal.new("1E100").to_s` should output `"1E100"` rather than  `"10000...000"` (100 zeros)

the expanded version is arguably not much use and generating it can become very expensive (sort of related to #7808).

In other langs, 
* Ruby and Python output decimals as strings in E notation, 
* JS does too when Number value is large

At least if `.to_s` continues to create a representation that is expanded, it should be documented if it happens by default.

I would be happy to contribute to the implementation of this should it be agreed upon!

(originally part of #9547)

### PR DESCRIPTION
**Pull request by HertzDevil**
Resolves #9578.

```crystal
# before
BigDecimal.new(Float32::MAX)          # => 340282350000000000000000000000000000000
BigDecimal.new(Float32::MIN_POSITIVE) # => 0.000000000000000000000000000000000000011754944

# after
BigDecimal.new(Float32::MAX)          # => 3.4028235e+38
BigDecimal.new(Float32::MIN_POSITIVE) # => 1.1754944e-38
```

The code here is very similar to #10632 and `Float::Printer#internal`. If both are merged we should think about refactoring them into a single implementation.

### CODE REVIEW

File: src/big/big_decimal.cr [hunk 1]
```diff
@@ -305,35 +305,76 @@ struct BigDecimal < Number
     in_scale(0)
   end
 
+  # BUG: a
+  # DEPRECATED: b
+  # FIXME: c
+  # NOTE: d
+  # OPTIMIZE: e
+  # TODO: f
+  # EXPERIMENTAL: g
```
Comment (Sija) [inline src/big/big_decimal.cr:314]: Debug leftovers?

```suggestion
```

File: src/big/big_decimal.cr [hunk 2]
```diff
@@ -312,28 +312,62 @@ struct BigDecimal < Number
...
-    s = @value.to_s
-    if @scale == 0
-      io << s
-      return
+    cstr = LibGMP.get_str(nil, 10, @value)
+    length = LibC.strlen(cstr)
+    buffer = Slice.new(cstr, length)
+
+    # add negative sign
+    if buffer[0]? == 45 # '-'
```
Comment (straight-shoota) [inline src/big/big_decimal.cr:320]:
```suggestion
    if buffer[0]? === '-'
```
I'm pretty sure this doen't add any significant overhead and saves you from annotating the character in a comment ;)

**Review by straight-shoota — APPROVED** [review]
This doesn't break any formal API, but it's still a significant change in behaviour. I you expect ˋ0ˋ to stringify as ˋ0ˋ, returning ˋ0.0ˋ breaks that. 

So... do we have to wait for 2.0 with this?

**Comment by HertzDevil** [comment]
Some additional clarification: unlike https://github.com/crystal-lang/crystal/pull/10632#issuecomment-885418088, fixed precision formatting is actually possible for `BigDecimal`s converted from `Float`s, but the current constructor doesn't work because it relies on `Float#to_s`. Instead the float must be decomposed manually:

```crystal
require "big"

struct BigDecimal
  def **(other : Int) : BigDecimal
    return (to_big_r ** other).to_big_d if other < 0 # #10892
    BigDecimal.new(@value ** other, @scale * other)
  end

  def self.exact(float : Float32)
    mantissa, exponent = Math.frexp(float)
    Math.ldexp(mantissa, Float32::MANT_DIGITS).to_i * 2.to_big_d ** (exponent - Float32::MANT_DIGITS)
  end

  def self.exact(float : Float64)
    mantissa, exponent = Math.frexp(float)
    Math.ldexp(mantissa, Float64::MANT_DIGITS).to_i * 2.to_big_d ** (exponent - Float64::MANT_DIGITS)
  end
end

# Before:
BigDecimal.exact(Float32::MAX)          # => 340282346638528859811704183484516925440
BigDecimal.new(Float32::MAX)            # => 340282350000000000000000000000000000000
BigDecimal.exact(Float32::MIN_POSITIVE) # => 0.0000000000000000000000000000000000000117549435082228750796873653722224567781866555677208752146087936
BigDecimal.new(Float32::MIN_POSITIVE)   # => 0.000000000000000000000000000000000000011754944

# After:
BigDecimal.exact(Float32::MAX)          # => 3.4028234663852885981170418348451692544e+38
BigDecimal.new(Float32::MAX)            # => 3.4028235e+38
BigDecimal.exact(Float32::MIN_POSITIVE) # => 1.17549435082228750796873653722224567781866555677208752146087936e-38
BigDecimal.new(Float32::MIN_POSITIVE)   # => 1.1754944e-38
```

Like the corresponding PR for `BigFloat`, this patch never changes the number of significant digits in the output of `BigDecimal#to_s`.

**Review by beta-ziliani — APPROVED** [review]
I don't think there were complains about pushing #10632, so I think we can follow suit here. Sorry Quinton that it took this long, and thanks! 🙇

File: src/big/big_decimal.cr (hunk 2 above)
Comment (sevk) [reply src/big/big_decimal.cr:451]: “===” different from ruby syntax ？

**Pull request merged**